### PR TITLE
Refactor runASTAnalysis functions to use class AstAnalyser

### PR DESCRIPTION
**Pull request by jean-michelet**
The aim of this PR is to make the project more flexible and allow users to add new behaviors more easily.

I've created an `AstAnalyser` class that implements the logic originally contained in `runASTAnalysis` and `runASTAnalysisOnFile`. The class can be extended and has a dependency on any EStree-compatible parser that respects the following abstraction:
```ts
interface SourceParser {
  parse(source: string, options: unknown): ESTree.Program;
}

interface AstAnalyser {
  constructor(parser: SourceParser): void;
  analyse: (str: string, options?: Omit<RuntimeOptions, "customParser">) => Report;
  analyzeFile(pathToFile: string, options?: Omit<RuntimeFileOptions, "customParser">): Promise<ReportOnFile>;
}
```
Example:
```js
const jsAnalyser = new AstAnalyser(new JsSourceParser());
const tsAnalyser = new AstAnalyser(new TsSourceParser());
```
I removed the  class `SourceParser` to favor ***composition over inheritance***. 
Responsibility for preparing source code (deleting html comments etc.) for analysis can be assigned to the new class `AstAnalyser`. It removes the need for this class from the workspaces, wich could be problematic to create the `TsSourceParser` ([214](https://github.com/NodeSecure/js-x-ray/issues/214)).

I kept the functions `runASTAnalysis` and `runASTAnalysisOnFile` so my additions extend the possibilities without breaking the existing. Users can pass a custom parser to these functions without having to completely modify their code:
```js
function runASTAnalysis(
  str,
  options = Object.create(null)
) {
  const {
    customParser = new JsSourceParser(),
    ...opts
  } = options;

  const analyser = new AstAnalyser(customParser);

  return analyser.analyse(str, opts);
}
```

I've moved the tests for these functions to the new `AstAnalyser` class and replaced them with integration tests verifying that the functions call the `AstAnalyser` methods with the right arguments.

### CODE REVIEW

File: types/api.d.ts [hunk 1]
```diff
@@ -76,5 +73,15 @@ type ReportOnFile = {
   warnings: Warning[];
 }
 
+interface SourceParser {
+  parse(source: string, options: unknown): ESTree.Program;
+}
+
+interface AstAnalyser {
```
Comment (fraxken) [inline types/api.d.ts]: declare class AstAnalyser plutôt non?

File: types/api.d.ts [hunk 2]
```diff
@@ -76,5 +73,15 @@ type ReportOnFile = {
   warnings: Warning[];
 }
 
+interface SourceParser {
```
Comment (fraxken) [inline types/api.d.ts:76]: declare class

File: src/AstAnalyser.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
+// Import Node.js Dependencies
+import fs from "fs/promises";
+import path from "path";
```
Comment (fraxken) [inline src/AstAnalyser.js]: You need to add `node:` prefix

File: src/AstAnalyser.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import isMinified from "is-minified-code";
+
+// Import Internal Dependencies
+import { SourceFile } from "./SourceFile.js";
+import { isOneLineExpressionExport } from "./utils/index.js";
+
+export class AstAnalyser {
+  /**
+   * @constructor
+   * @param { Parser } parser
```
Comment (fraxken) [inline src/AstAnalyser.js]: What is Parser type?

**Review by fraxken — CHANGES_REQUESTED** [review]

File: src/AstAnalyser.js [hunk 5]
```diff
@@ -1,6 +1,6 @@
 // Import Node.js Dependencies
-import fs from "fs/promises";
-import path from "path";
+import fs from "node:fs/promises";
+import path from "node:path";
```
Comment (jean-michelet) [inline src/AstAnalyser.js:3]: Ooops @fraxken 

Added this fix in the wrong commit bu it's there

**Review by fraxken — APPROVED** [review]

**Review by PierreDemailly — APPROVED** [review]

**Comment by fraxken** [comment]
@allcontributors please add @jean-michelet for code, test, doc

**Pull request merged**